xapian-applications/omega/query.cc

```diff
@@ -651,25 +652,38 @@ html_strip(const string &str)
     return res;
 }
 
-// FIXME split list into hash or map and use that rather than linear lookup?
-static int word_in_list(const string& word, const string& list)
-{
-    string::size_type split = 0, split2;
-    int count = 0;
-    while ((split2 = list.find('\t', split)) != string::npos) {
-	if (word.size() == split2 - split) {
-	    if (memcmp(word.data(), list.data() + split, word.size()) == 0)
-		return count;
+static unordered_map<string, int> word_to_occurrence;
+
+class WordList {
+    string list;
+  public:
+    WordList() : list("") { }
+
+    void build_word_map(const string& list2) {
+	// Don't build map again if passed list of terms is same as before.
+	if (list == list2) return;
```

Comment:
> This optimisation still won't work as the code currently is - you create a new WordList object each time html_highlight() is called, so list will get reset each time and (assuming list2 isn't empty) the mapping always gets rebuilt.

I see the mistake made here. As you suggested, `list` must be made static and also the map be encapsulated inside the class.

> Stylistically, list2 seems ugly - it would be clearer to call the parameter list and the member variable something like prev_list.

Yes, agreed. Will make this change as well.